### ISSUE TITLE
Publishing rate can temporarily exceed the requested rate

### ISSUE THREAD

**Issue by mkuratczyk** [rabbitmq/rabbitmq-perf-test#386]
Here's one example (note: there is no `-c` here) - why did perf-test send 84msgs?

```
$ perf-test -x 5 -y 0 -vr 10:10 -vr 0:10
id: test-105343-863, time: 1.015s, sent: 45 msg/s
id: test-105343-863, time: 2.015s, sent: 55 msg/s
id: test-105343-863, time: 3.015s, sent: 55 msg/s
id: test-105343-863, time: 4.015s, sent: 55 msg/s
id: test-105343-863, time: 5.016s, sent: 54 msg/s
id: test-105343-863, time: 6.016s, sent: 55 msg/s
id: test-105343-863, time: 7.017s, sent: 54 msg/s
id: test-105343-863, time: 8.017s, sent: 55 msg/s
id: test-105343-863, time: 9.017s, sent: 55 msg/s
id: test-105343-863, time: 10.017s, sent: 55 msg/s
id: test-105343-863, time: 20.119s, sent: 0.49 msg/s
id: test-105343-863, time: 21.119s, sent: 84 msg/s
id: test-105343-863, time: 22.119s, sent: 55 msg/s
id: test-105343-863, time: 23.119s, sent: 55 msg/s
id: test-105343-863, time: 24.120s, sent: 54 msg/s
```

Here's another example (I'm almost sure many of the flags are unnecessary to reproduce the problem, but sharing an actual thing I ran into; the slow down was caused by a server-side slow-down): 
```
$ perf-test -x 5 -y 0 -qp q-%d -qpf 1 -qpt 5 -c 1000 -r 10000 -ad false -s 100 -mp deliveryMode=1 -qa x-max-length=5000000,x-queue-version=2 -ms

id: test-095750-804, time: 245.087s, sent: 50004 msg/s, confirmed: 49980 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 0/0/1/3/7 ms
id: test-095750-804, time: 246.087s, sent: 50001 msg/s, confirmed: 50055 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 0/0/1/1/7 ms
id: test-095750-804, time: 247.087s, sent: 49989 msg/s, confirmed: 49990 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 0/0/1/2/7 ms
id: test-095750-804, time: 248.087s, sent: 50010 msg/s, confirmed: 49917 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 0/0/1/1/6 ms
id: test-095750-804, time: 249.087s, sent: 49990 msg/s, confirmed: 50063 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 0/0/1/2/6 ms
id: test-095750-804, time: 250.118s, sent: 29548 msg/s, confirmed: 24733 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 0/0/1/153/329 ms
id: test-095750-804, time: 251.121s, sent: 1659 msg/s, confirmed: 1652 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 396/1063/1346/1544/1578 ms
id: test-095750-804, time: 252.132s, sent: 2785 msg/s, confirmed: 2778 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 997/1808/1980/2087/2324 ms
id: test-095750-804, time: 253.132s, sent: 18111 msg/s, confirmed: 18119 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 27/95/886/2462/3051 ms
id: test-095750-804, time: 254.132s, sent: 70800 msg/s, confirmed: 70904 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 20/70/87/114/127 ms
id: test-095750-804, time: 255.132s, sent: 72573 msg/s, confirmed: 72496 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 25/67/82/105/116 ms
id: test-095750-804, time: 256.132s, sent: 71256 msg/s, confirmed: 71246 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 24/69/83/106/118 ms
id: test-095750-804, time: 257.132s, sent: 68884 msg/s, confirmed: 68896 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 24/72/89/113/128 ms
id: test-095750-804, time: 258.133s, sent: 71060 msg/s, confirmed: 71037 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 21/68/84/106/126 ms
id: test-095750-804, time: 259.133s, sent: 70432 msg/s, confirmed: 70424 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 24/70/83/108/121 ms
id: test-095750-804, time: 260.133s, sent: 66791 msg/s, confirmed: 68406 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 3/63/80/107/122 ms
id: test-095750-804, time: 261.133s, sent: 57066 msg/s, confirmed: 59254 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 0/40/76/109/132 ms
id: test-095750-804, time: 262.133s, sent: 50301 msg/s, confirmed: 51020 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 0/9/17/38/80 ms
id: test-095750-804, time: 263.133s, sent: 50010 msg/s, confirmed: 49884 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 1/8/13/23/34 ms
id: test-095750-804, time: 264.133s, sent: 50009 msg/s, confirmed: 50544 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 0/4/10/21/26 ms
id: test-095750-804, time: 265.133s, sent: 49990 msg/s, confirmed: 50047 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 0/0/1/1/3 ms
id: test-095750-804, time: 266.133s, sent: 50000 msg/s, confirmed: 50000 msg/s, nacked: 0 msg/s, min/median/75th/95th/99th confirm latency: 0/0/0/1/1 ms
```

And yet another example. In this case, the slow down just before the spike was relatively minor, although long-lasting. I'm not even sure how perf-test noticed that the slow-down is no more. Perhaps it received exactly 50000 confirms (or more than that), which triggered it to send more than the expected 50k?
![Screenshot 2022-09-22 at 12 22 31](https://user-images.githubusercontent.com/9566114/191748475-ec09d081-1af2-4b19-8b77-639f0620d4a4.png)

**Comment by acogoluegnes** [comment]
Tip: execute `eprof:start(). eprof:start_profiling([<0.6686.3>]). timer:sleep(10000). eprof:stop_profiling(). eprof:analyze().` in a shell (change the PID to a queue PID) to disrupt the system (second example).

**Comment by acogoluegnes** [comment]
Fixed (partially) in #405.
